apps/opentelemetry_experimental/src/otel_log_handler.erl

```diff
@@ -12,227 +12,457 @@
-handle_event({call, From}, {filter_handler, Config}, Data) ->
-    {keep_state, Data, [{reply, From, Config}]};
-handle_event({call, From}, {filter_config, Config}, Data) ->
-    {keep_state, Data, [{reply, From, Config}]};
-handle_event({call, _From}, _Msg, _Data) ->
-    keep_state_and_data;
-handle_event(cast, {log, Scope, LogEvent}, Data=#data{batch=Logs}) ->
-    {keep_state, Data#data{batch=maps:update_with(Scope, fun(V) ->
-                                                                  [LogEvent | V]
-                                                          end, [LogEvent], Logs)}};
-handle_event(_, _, _) ->
-    keep_state_and_data.
+exporting(enter, _OldState, #data{exporter=undefined,
+                                  atomic_ref=AtomicRef,
+                                  tables=Tabs}) ->
+    %% exporter still undefined, go back to idle.
+    %% First empty the table and disable the processor so no more log events are added.
+    %% We wait until the attempt to export to disable so we don't lose log events
+    %% on startup but disable once it is clear the exporter isn't being set
+    clear_table_and_disable(AtomicRef, Tabs),
 
-%%
+    %% use state timeout to transition to `idle' since we can't set a
+    %% new state in an `enter' handler
+    {keep_state_and_data, [{state_timeout, 0, no_exporter}]};
+exporting(enter, _OldState, Data=#data{atomic_ref=AtomicRef,
+                                       tables=Tabs,
+                                       max_queue_size=MaxSize,
+                                       exporting_timeout_ms=ExportingTimeout,
+                                       scheduled_delay_ms=SendInterval}) ->
+    CurrentTab = ?current_tab(AtomicRef),
+    case ?get_available(AtomicRef, CurrentTab) of
+        %% No events yet, maximum available capacity, nothing to export
+        MaxSize ->
+            %% The other table may contain residual (late) writes not exported during
+            %% the previous run.
+            %% If current table is not empty,
+            %% we don't need to check the size of the previous (currently disabled) table,
+            %% since we will switch to it after this exporter run.
+            %% However, of current table remains empty for the long time, no export and table switch
+            %% will be triggered, and any residual late log events in the previous table will be left dangling.
+            maybe_export_other_table(CurrentTab, Data);
+        _ ->
+            RunnerPid = export_logs(CurrentTab, Data),
+            {keep_state,
+             Data#data{runner_pid=RunnerPid, handed_off_table=?tab_name(CurrentTab, Tabs)},
+             [{state_timeout, ExportingTimeout, exporting_timeout},
+              ?export_timeout(SendInterval)]}
+    end;
+
+exporting(state_timeout, no_exporter, Data) ->
+    {next_state, idle, Data};
+exporting(state_timeout, empty_table, Data) ->
+    {next_state, idle, Data};
+
+exporting(state_timeout, exporting_timeout, Data) ->
```

Comment:
There does not seem to be a cancellation issued to the exporter? then this is probably not the best strategy.

* if lower stack is jammed, the best option at the higher level is to either propagate the backpressure to its higher level.
* if current layer is a sink (no backpressure possible), then the second-best option is continue waiting for whatever had been sent, and drop new incoming calls.

going back to idel and trigger yet another batch later will probably make it jam even worse ?

Reply:
Yes, it simply kills the current exporter forcefully, we don't know if the exporter/receiving end will keep processing this batch request. 
Current overload protection only guarantees that next batch won't exceed `max_queue_size` and the exporter will be given at least `exporting_timeout` to cope with that batch before it can receive a next one. 
I think we need to set a deadline in exporter's GRPC client (probably slightly lower than exporting_timeout). If the receiving end is a well-behaved GRPC server, it should "stop any ongoing computation and end its side of the stream"
https://grpc.io/docs/guides/cancellation/#overview